### DISCUSSION TITLE
How to make installation easier?

### DISCUSSION THREAD

**Discussion by markseuffert** [datenstrom/community#187] [Ask a question] [answered]
I would like to make the installation easier. Since last release there's an installer which guides you trough creating user accounts, but setting up a complete website takes too much time. You have to copy plugin files into different places, read multiple installation documentation and figure out how to do updates without breaking stuff. All this gets in the way of what people want to do.

What can be done to make installation easier?

**Comment by nibreh** [comment]
Some ideas, just like that, don't know if it's possible or simple: 

-Create a system on the official website to download last yellow with selected plugins: Users can select plugins, language, etc he wants and the .zip file is generated with the features included. Then just install and create user online.

-Notification online when there is a new official release, with a link on how to update.

I'll think about ideas, I have to go ;)

**Comment by schulle4u** [comment]
> -Create a system on the official website to download last yellow with selected plugins: Users can select plugins, language, etc he wants and the .zip file is generated with the features included. Then just install and create user online.

An example for such a system is [the dokuwiki download
page](http://download.dokuwiki.org)

Github's email interface sucks. :/

**Comment by BsNoSi** [comment]
For me the installation is perfect. I copy required files — done. Same with updates: if you know, what to do (and what you are doing), it is as simple as installing.  If you want to update more than one installation: Unpack current version, delete all unwanted files (especially config,…) pack again and unzip on server. Done.

**Challenges (in my opinion)**
- Adding users 
  - have not tested new release — is it possible, to reuse the user-setup locally after install?
    - **UPDATE:** Tested current/new version with option „create new user“ — nice!
- Filemanager
  - Currently, only users with FTP-Access have ability to add media. This is a major obstacle against general usage. 
- If these two are available in the frontend:
  - A simple separation between administrator and (other) users.
    - Allowing/disallowing "add/modifiy user"
    - Allowing/disallowing "upload media", "update media“, "delete media"
    - Optional: Defining „writing area“ for users (folder, they can edit).

**Missing (in my opinion)**

A complete "How to setup / update Yellow" document. Maybe I will find some time this week and try to write one. If, I will send it to Mark.

**Comment by markseuffert** [comment]
@wunderfeyd @nasendackel @hysonlee @Zauberweib What do you think, what's your experience?

**Comment by hysonlee** [comment]
From my experience , the complicated thing is to update to new core when there with my customized plugs or snippets, 500 Error is inevitable.

If Yellow can handle the dependency or show the bug location automatic , Will good for new user and can save much time then.

**Comment by wunderfeyd** [comment]
Since a yellow installation is often heavily modified to meet the users wishes updates are hard to create. It will only successful if no templates and/or snippets are modified or created. Maybe a check for file hashes or new files would help here to make the update process more "stable" (at least the user will get the message "Please check the installation after the applied update).

From my point of view the installation is easy enough for people who are working with computers a long time or have a deeper understanding of the computer background. (Although sometimes when reading the new issues it seems a web front end for the CLI to yellow would be nice. But that's something for the far future maybe.)

For new users or one time users it might be complicated. As long everything is covered within the installation documents it could work. Yellow one time users (or even people who want to compare with other systems in this category) are not willing to read the documentation about styling, adding/setup plugins and maintaining compatibility with possible updates.

So in my option to catch the new and one time users, it would be fine to add a dialog in the webinterface to add/remove plugins and themes/styles. This dialog has to be the first thing they see after unzipping the base yellow installation. Maybe one could create customized zip files before installing yellow. Like...
- What's your webserver (maybe with "I don't know" selection)?
- What's your favorite layout? (with preview)
- What kind of installation do you need (blog/wiki/mixed)? (with preview)
- Which extras do you need? Images/Youtube/Sidebar? (with partial preview)
- Favorite directory structure (maybe depending on the the kind of installtion)?
- Type of modification? (Direct (dropbox/ftp)/Webinterface)
- User and passwords.

I'm not a user of the category so I've trouble to put myself into this position. :\ If I should be correct with my thinking it would be a big amount of work. So the question is: Is it useful? What kind of users one tries to reach.

**Comment by Zauberweib** [comment]
create an account could be better. for one-user: maybe on installing just ask about. and if theres more accounts needed - make an extra plugin for it? 

and for plugins you have to mark-copy-paste the php-file(s). i.e https://github.com/datenstrom/yellow-plugins/tree/master/toc - here I wish i could right-click on toc.php and save it where I need. but that doesn't work. 

please do _not_ make some backend/dashboard/trallala - I love yellow cause its so simple. In former (not further, former!) times I used "kirby", too. but as they came along with their "panel" I found it only horrible. don't do the same mistake. Making something good better stops being something good. 

thats my ideas so far...

and maybe some full-package-installation would be nice (choose: basic / all-in ;) )

**Comment by ghost** [comment]
I think Installation is very easy, just copy everything to your server. It is not more difficult then installing wordpress or other cms. It is even easier, because you don't have to set up a database. 

Updating could be easier, may release some snapshots in the future, like monthly builds that only replaces stuff, but keep the config, the template etc....

Just keep yellow that simple. I can manage my blog with only a terminal, ssh and an editor. Great.

**Comment by BsNoSi** [comment]
I think, the question is, who is targeted with the update. If we want to support non-experienced users, it is a hard challenge, to have simple and "working under any circumstances" install/update.

Install is in current state really simple, because you have to copy files, run and some questions (user) complete it. Great.

The real challenge is updating. I follow wunderfeyd in his/her (?) argumentation, that updating is time consuming. I noticed, that my idea (take current files and overwrite) does not grant success. Especially changes in config.ini or system files can create a bug tracing session, why things do no more behave as expected.

I am no fan of "everything from the interface". Typically, a website is not modified every week. There is no need for a install/modify interface, that is only occasionally required.

If there is a demo area (given with the demo version) demonstrating the plugins, a.s.o., it is simple to decide, what you need/want. Then copying the plugin into the respective folder is a simple task. No interface for that is a security feature: only admins can do that ;)

User account setup has significantly improved. Very nice job Mark!

Maybe a better documented update can help. If a new update is available, 
- short notes, which changes my influence an existing installation,
- a notice, if updating core requires updated plugins.

I am not shure, if this is "best practice", but I [download yello master.zip](https://github.com/datenstrom/yellow/archive/master.zip) from github, [yellow-plugins.zip](https://github.com/datenstrom/yellow-plugins/archive/master.zip) and use it for updating a testing area (backup of life system). If anything works as expected, I set this to life, otherwise, I start bug hunt. I leave theme updating, because I do not use them as given. This can cause some disturbance, but this is my personal challenge.

**Comment by redtide** [comment]
I think a plugin and theme installer could be nice to have from interface.
About the core update I don't know, it could be more tricky, I still don't know much how yellow works (e.g. I saw some update.ini files in some repository but I have no idea what are those for).
Anyway I will be available to contribute on this.

**Comment by markseuffert** [comment]
I am working on this now. 

What's already working: Installing a plugin is downloading and copying **ONE** zip file. 

As a developer you should add an `update.ini` file to your GitHub repositories, which is later has been renamed to `extension.ini`. This takes care of installing files and will also be used by software updates later on. The following file operations are supported in plugins/themes:

create = create if not exists
update = overwrite if exists
delete = delete if exists
careful = only if not modified
optional = only if new software installation

Let me know if you have questions or more suggestions.

**Comment by Zauberweib** [comment]
Maybe some silly question: any possibility to get ask bevor overwriting existing files? 
Beside of this I think its cool :)

**Comment by markseuffert** [comment]
What happens during software update is that only files under version control are overwritten. So there's no question for individual files. The current implementation is quite careful to not touch any customizations that users have done in templates and snippets. In the best case, you press a button and let Yellow do the work for you. In the worst case, you can find files that have been overwritten or deleted in your `system/trash` folder. Nothing can get lost. Let's see how well that works?

**Comment by markseuffert** [comment]
The software update is available now.

Here's what's has been changed:
- use commandline to update software with little effort
- use webinterface to “Check for updates”, shows if updates are available
- use webinterface to update software with a single click, available for webmaster only
- use webinterface for creation/deletion of content folders, integrated into page editor
- added image plugin to default installation, should make life easier for new users
- checked that nothing gets lost, deleted files will go into your `system/trash` folder

I forgot to mention that updating plugins does also work like installing plugins: Updating a plugin is downloading and copying ONE zip file. For those who prefer a manual software update.

Let me know how this works for you.

**Comment by markseuffert** [comment]
I think I came across another installation problem.

If you want to use Yellow as static site generator, you have to configure your server settings. There are three different server settings, which makes sense from a technical perspective but puts a hurdle in the way of new users. Finding out the correct settings isn't easy, as I have seen from support emails. 

Perhaps there's a better solution, for example using the website URL? More problems I should look at?

**Comment by ifrx** [comment]
For me (a user, not a programmer), Yellow is such a pleasant surprise after years looking (occasionally) for a clearly structured, markdown-oriented minimalist flat-file website system.
Yellow is encouragingly slim, it is wellstructured, clearly named, and commented, it is surprisingly comprehensible, it is just text (md, html, css) and php. No heaps of fancy helper languages on top of it.
To operate, all it needs is ssh, a file manager, and a text editor.

It largely does what it says. And if this is not the case, it is just a bit of missing information. I would say, as long as Yellow stays lean and clearly structured, and all plugins follow the same guidelines, it is probably always much more reliable to unzip and to adjust manually with the help of an up-to-date doc, than to have to hope that a suitable magic routine is in place.

However, I can see your point about automatic setups and encouraging new users with a love for installers.
What about turning the installer into a specific plugin (like update.php), so that people could decide if they want to use it or would rather go manual or console?
What about two download options: Yellow (minimal manual setup) or Yellowmat (setup routine)?

**Comment by markseuffert** [comment]
Thanks for your feedback, this is much appreciated.

How it works at the moment: The installer helps users with basic configuration. However this is fully optional. It takes one click to skip the installer and it is never shown at the command line. You can turn off the installer by setting `InstallationMode: 0` in the system settings.

About your suggestions, if more people would like to change the installer we can do it.

**Comment by markseuffert** [comment, marked as answer]
The software update is available now, second attempt.

Here's what has been changed:

* software update has been rewritten, to ignore modified files
* webmaster can now force a software update to overwrite files
* improved themes to keep modified CSS files, new directory structure
* improved configuration for static site generator, new URL configuration
* improved installation in multi language mode, detects content structure
* tested that update (from previous version) is as smooth as possible, converts old files
* tested that update takes care of existing themes, imports custom CSS files
* added plugin event to detect when software has been updated

A short summary of what's possible now:

* installer helps users with creating user accounts and basic configuration
* installing features is downloading and copying zip files
* updating website with little effort, use web browser or command line

The installation and maintenance of a website has become much easier. Hopefully this lowers the entry barrier and more people try out Yellow. There's an sentence that was always important to me: "If you have to read the instructions first, then we have done something wrong". Unfortunately, you had to read many instructions. Finally it's possible, you can install a small website in a few seconds. 

Thanks for all your feedback. Let me know how this works for you.

**Comment by BsNoSi** [comment]
I have to admit, that I checked twice, if update (0.6.7 → 0.6.8) has really happened. I did it in a live system (buoa.de) – I trust in your work ;)

What can I say: **Really good job – thank you!**
